connector/topologyconnector/internal/deduplicator.go

```diff
@@ -0,0 +1,409 @@
+//   - Cache eviction only affects memory usage; it does not change the semantic
+//     correctness of deduplication. Evicted entries will be treated as new on the
+//     next invocation.
+func (d *TopologyDeduplicator) ShouldSend(
+	mappingIdentifier string,
+	signal settings.OtelInputSignal,
+	evalCtx *ExpressionEvalContext,
+	expressionRef *types.ExpressionRefSummary,
+	mappingTTL time.Duration,
+) bool {
+	if !d.enabled || expressionRef == nil {
+		return true
+	}
+
+	if d.cacheTTL > 0 && d.cacheTTL < mappingTTL { // cacheTTL == 0 means no eviction
+		d.logger.Warn(
+			"Dedup cache TTL is smaller than mapping TTL; entries may be evicted prematurely.",
+			zap.Duration("cacheTTL", d.cacheTTL),
+			zap.String("mappingIdentifier", mappingIdentifier),
+			zap.Duration("mappingTTL", mappingTTL),
+		)
+	}
+
+	now := time.Now()
+	refreshBefore := time.Duration(float64(mappingTTL) * d.refreshFraction)
+
+	// Generate deterministic key from the projection of signal inputs.
+	// This serves as both the cache key and content identifier.
+	// Any change in the projection results in a different key, naturally
+	// creating a new cache entry for what is effectively a different series.
+	key := d.hasher.ProjectionHash(mappingIdentifier, signal, evalCtx, expressionRef)
+
+	d.logger.Debug(
+		"Deduplication key", zap.String("mappingIdentifier", mappingIdentifier), zap.String("key", key),
+	)
+
+	e, ok := d.cache.Get(key)
+	if !ok {
+		// New series (or content changed creating new key) -> send and record
+		d.cache.Add(key, dedupEntry{lastSent: now})
+		return true
+	}
+
+	// Existing series with same content -> only refresh according to time policy
+	if now.Sub(e.lastSent) >= refreshBefore {
+		d.cache.Add(key, dedupEntry{lastSent: now})
+		return true
+	}
+
+	// Skip sending - content unchanged and not time to refresh
+	return false
+}
+
+type DedupHasher interface {
+	// ProjectionHash generates a stable identifier based on structure, not content
+	ProjectionHash(
+		mappingIdentifier string,
+		signal settings.OtelInputSignal,
+		ctx *ExpressionEvalContext,
+		ref *types.ExpressionRefSummary,
+	) string
+}
+
+type SignalMappingProjectionHasher struct{}
+
+// ProjectionHash computes a stable hash for the subset of signal data to deduplicate on.
+// It always includes the full resource map and only the referenced vars/attributes for the rest.
+// If ref is nil, fall back to hashing only mapping+signal to avoid false positives.
+func (s *SignalMappingProjectionHasher) ProjectionHash(
+	mappingIdentifier string,
+	signal settings.OtelInputSignal,
+	ctx *ExpressionEvalContext,
+	ref *types.ExpressionRefSummary,
+) string {
+	h := xxhash.New()
+
+	// Domain separator and identifiers
+	mustWrite(h, []byte("projection_v1\x00"))
+	mustWrite(h, []byte(mappingIdentifier))
+	mustWrite(h, separator)
+	mustWrite(h, []byte(signal))
+	mustWrite(h, separator)
+
+	// Resource (always include full resource)
+	if ctx.Resource != nil {
+		mustWrite(h, []byte("resource:"))
```

Comment:
To improve efficiency a bit we can make these static strings much shorter so they are only 1 character or 1 byte.